include/rmm/detail/dynamic_load_runtime.hpp

```diff
@@ -0,0 +1,129 @@
+      if (!ptr) { ptr = ::dlopen(libname.c_str(), RTLD_LAZY); }
+      if (ptr) { return ptr; }
+
+      RMM_FAIL("Unable to dlopen cudart");
+    };
+    static std::unique_ptr<void, decltype(close_cudart)> cudart_handle{open_cudart(), close_cudart};
+    return cudart_handle.get();
+  }
+
+  template <typename... Args>
+  using funcion_sig = std::add_pointer_t<cudaError_t(Args...)>;
+
+  template <typename signature>
+  static signature function(const char* func_name)
+  {
+    auto* runtime = get_cuda_runtime_handle();
+    auto* handle  = ::dlsym(runtime, func_name);
+    if (!handle) { return nullptr; }
+    auto* function_ptr = reinterpret_cast<signature>(handle);
+    return function_ptr;
+  }
+};
+
+#if CUDART_VERSION >= 11020  // 11.2 introduced cudaMallocAsync
+/**
+ * @brief `async_alloc` bind to the Stream Ordered Memory Allocator functions
+ * at runtime.
+ *
+ * This allows us rmm users to compile/link against CUDA 11.2+ and run with
+ * < CUDA 11.2 runtime as these functions are found at call time
+ */
+struct async_alloc {
+  static bool is_supported()
+  {
+#if defined(RMM_STATIC_CUDART)
+    static bool runtime_supports_pool = (CUDART_VERSION >= 11020);
+#else
+    static bool runtime_supports_pool =
+      dynamic_load_runtime::function<dynamic_load_runtime::funcion_sig<void*, cudaStream_t>>(
+        "cudaFreeAsync") != nullptr;
+#endif
+
+    static auto driver_supports_pool{[] {
+      int cuda_pool_supported{};
+      auto result = cudaDeviceGetAttribute(&cuda_pool_supported,
+                                           cudaDevAttrMemoryPoolsSupported,
+                                           rmm::detail::current_device().value());
+      return result == cudaSuccess and cuda_pool_supported == 1;
+    }()};
+    return runtime_supports_pool and driver_supports_pool;
+  }
+
+#if defined(RMM_STATIC_CUDART)
+#define RMM_SYNC_ALLOC_WRAPPER(name, signature)              \
+  template <typename... Args>                                \
+  static cudaError_t name(Args... args)                      \
+  {                                                          \
+    static const auto func = static_cast<signature>(::name); \
+    return func(args...);                                    \
+  }
+#else
+#define RMM_SYNC_ALLOC_WRAPPER(name, signature)                                \
+  template <typename... Args>                                                  \
+  static cudaError_t name(Args... args)                                        \
+  {                                                                            \
+    static const auto func = dynamic_load_runtime::function<signature>(#name); \
+    return func(args...);                                                      \
+  }
+#endif
+
+  template <typename... Args>
+  using cudart_sig = dynamic_load_runtime::funcion_sig<Args...>;
+
+  using cudaMemPoolCreate_sig = cudart_sig<cudaMemPool_t*, const cudaMemPoolProps*>;
```

Comment:
Something that just occurred to me. Having the symbols like `cudaMemPool_t` won't cause an issue, will it?

I assume that since it's not actually _using_ these symbols it doesn't need to bind them at load time? i.e., these don't need to be complete types (so we could forward declare them if we wanted to?).

Reply:
looking through the nm output i don't see any issue as no functions that use that type are marked undefined. 

In testing with the latest cuda driver, and 11.0 runtime I see no issues with the tests and undefined symbols ( and the tests behave as expected ).